[WIP] drivers: mfd: add support for stm32 timers

maxvankessel opened the pull request:
So, I was a bit upset due to extensibility of the timer driver for the stm32 within Zephyr.
I opened up an enhancement issue #20262. And started digging around how Linux does this.

So I got inspired by by the MFD driver for stm32-timers, see [stm32-timers](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/mfd/stm32-timers.txt?h=v5.4-rc5).

I'm using the stm32f4 series as a motorcontroller (with hall sensors and field oriented control). And the timer can do a pretty good job here. But I want to use the timer as a base for other modules, like hall sensors and 3 phase generation. So I got it to work, and changed it only for the F4 series (all is based on the original drive, but I moved from the bulky HAL driver to to LL driver).

Also I moved from the DTS fixups to the instance based implementation. But it still needs more cleanup.

So I would like to hear from you, if it's useful I will cleanup things further so it can be merged.

**TODO's**
- Fix DTS files for complete stm32 series
- Cleanup dts_fixups
- Test on more platforms

erwango commented:
@maxvankessel, thanks for this contribution. Looks pretty nice with interesting bonuses like transition to LL API.
Since we're getting closer to V2.1 end of merge window. I'm tagging V2.2 as a merge target.
Also it is likely that closer review won't happen before V2.1 release.

erwango commented:
@avisconti , FYI, IIRC, you need a timer driver for some application

avisconti commented:
> @avisconti , FYI, IIRC, you need a timer driver for some application

You remember correctly. It was the #17140 PR, which is frozen for the moment as there was some
points to be clarified (timer driver was one of the points).
I may come back again to this PR in the next future.
Thanks!!

erwango commented on drivers/mfd/stm32_timers.c:
```diff
@@ -0,0 +1,220 @@
+
+		LL_TIM_SetPrescaler(tim, cfg->prescaler - 1);
+
+		/* TODO find a more consistent solution for all soc's */
```
Are there stm32 series on which this is not working ?

maxvankessel replied:
The comment corresponds to `TIM_CR1_DIR_Pos`. Maybe you could give some comments on it.

My opinion: `TIM_CR1_DIR_Pos` is not very portable over several series. It might be different on other devices. (I've seen this on the ADC)

Same goes for: `TIM_CR1_CMS_Pos`, `TIM_SMCR_SMS_Pos`, `TIM_SMCR_TS_Pos`, `TIM_CR2_MMS_Pos`

erwango commented on drivers/mfd/stm32_timers.c:
```diff
@@ -0,0 +1,220 @@
+		LL_TIM_SetPrescaler(tim, cfg->prescaler - 1);
+
+		/* TODO find a more consistent solution for all soc's */
+		if (cfg->align_mode == 0) {
```
Can you use enum dts bindings to get this more readable ?

maxvankessel replied:
Done

erwango commented on drivers/mfd/stm32_timers.c:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	/* Timer clock depends on APB prescaler */
+	clock_control_get_rate(data->clock,
```
Please check return value.

maxvankessel replied:
Done

erwango commented on drivers/mfd/stm32_timers.c:
```diff
@@ -0,0 +1,220 @@
+
+#ifdef DT_INST_5_ST_STM32_TIMERS
+TIMER_DEVICE_INIT(5);
+#endif
```
A reason to limit to 5 the number of enable instances ?

maxvankessel replied:
No, just had to start somewhere.

erwango commented on drivers/mfd/stm32_timers.h:
```diff
@@ -0,0 +1,82 @@
+ * @return negative errno code
+ */
+typedef int (*stm32_timers_get_cycles_per_sec_t)(struct device *dev,
+							u64_t * cycles);
```
Please fix alignment. Apply everywhere.

maxvankessel replied:
What kind of alignment do you propose, coding style is not very clear on this subject.

erwango replied:
Second line argument should start bellow first line argument

erwango commented on drivers/mfd/stm32_timers.h:
```diff
@@ -0,0 +1,82 @@
+};
+
+
+static inline void stm32_timers_enable(struct device *dev)
```
Are these available to user space as well?
If yes, you should add __syscall function similarly to what is done in other API.

maxvankessel replied:
On creation my intention was, that this driver can only be called by system drivers. 
But I use it within a user-space driver, so will change this.

erwango replied:
Makes sense, fine.

erwango commented on dts/bindings/timer/st,stm32-timers.yaml:
```diff
@@ -1,18 +1,65 @@
+    st,align-mode:
+      type: int
+      default: 0
+      description: Edge align or center align(1, 2, etc)
```
Can't we use enum here also ?

maxvankessel replied:
Done

erwango reviewed (changes requested):
I'm not a timer driver user, so I can't comment much on functional side of the driver.
I'l try to get more reviewers on that side.

Though, I've made a first review in term of integration in zephyr driver model. And this is already in good shape on that side.  Few comments though:
- To match other divers naming convention, please rename timer driver to timer_ll_stm32.c (or mfd_timer_...). 
- Also rename functions with same name space timer_stm32_enable, ...
- Split your PR in few commits. I think that at least the mfd introduction might be reviewed in wider audience, even tought it is stm32 specific (keep one PR for now though)

maxvankessel commented on drivers/mfd/stm32_timers.c:
```diff
@@ -0,0 +1,220 @@
+	u32_t prescaler;
+};
+
+static inline void __tim_stm32_get_clock(struct device *dev)
```
So I got this from the original driver, and I'm not 100% sure what it is what you want in return for it?

erwango replied:
> So I got this from the original driver,

This would have to be fixed. `tim_stm32_get_clock` would work.

jmbrederveld commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -110,216 +92,145 @@ static int pwm_stm32_pin_set(struct device *dev, u32_t pwm,
+	}
 
-	HAL_TIM_PWM_ConfigChannel(TimerHandle, &sConfig, channel);
+	LL_TIM_CC_EnableChannel(tim, timer_channels[pwm - 1]);
```
Enable "Main Output Enable" for timer1 & timer8

```
if (IS_TIM_BREAK_INSTANCE(tim)) {
	LL_TIM_EnableAllOutputs(tim);
}
```

erwango commented:
@maxvankessel, I'm starting having a deeper look.

To see real interest of the mfd, it would be nice to have an (even light) implementation of another driver based on this mfd_timer. Current in-tree potential user would be counter(timer).
Would you have time to work on a mfd based stm32 counter_timer ?

Also, to ease the review, it would be nice to split the PR in commits (spliting based on subystem would be a good start for now). This would be anyway a pre-requisite if we want to get this merged.

maxvankessel commented:
@erwango So I've split the commit, removed pwm contents from the mfd commit.

What do you mean with "counter"?

> Current in-tree potential user would be counter(timer).

 Quadrature decoder? Or a capture interface? What is it you want to count?

maxvankessel committed:
drivers: mfd: add support for stm32 mfd timers

Added multi functional device driver.
Added support for new timer base (mfd stm32 timers),
which aims to be more versatile.

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

erwango commented:
> @erwango So I've split the commit, removed pwm contents from the mfd commit.

Thanks for having taken the point into account, but I didn't meant to remove the PWM impact purely for the PR. Splitting the PR into several commits is generally a good way to ease the review  process (then maybe we split the PR into several PRs to speed up the merge of large PRs, but we're not there yet).

> What do you mean with "counter"?
Quadrature decoder? Or a capture interface? What is it you want to count?

`include/drivers/counter.h` is zephyr public API for counter and timer drivers.
You can find it's documentation here: https://docs.zephyrproject.org/latest/reference/peripherals/counter.html

Since interest of mfd is to serve multiple drivers, it would be valuable to demonstrate it within this change. This is why it would be valuable in your PR to have both pwm and counter demonstrated as timer-mfd users.

galak commented:
I don't see any particular issue with an MFD approach, just that the MFD subsystem should probably handle / provide tracking of the "child" devices.

maxvankessel commented:
I'm working on it. Something else came up.

henrikbrixandersen commented:
In the Linux kernel, MFD is more aimed at devices which are able to fulfill multiple different device APIs at once (like an RTC which also provides an EEPROM). As I read this, the STM32 timers are able to fulfill only _one_ out of many possible device APIs/roles at a time?

maxvankessel commented:
@henrikbrixandersen I hear what you are saying, it's more like a base for all timer based drivers.
The ST timer is a base for multiple drivers, but I'm not sure if it could do more than one function at a time. It could trigger the DMA.

I have made a quadrature decoder, and I realised it was a near copy of the PWM driver. 
Then I implemented a capture interface, again a near copy of the PWM driver. 
So I needed a base driver, because there were more features I needed, e.g. HALL driver and PWM outputs for motor control. A timer could also be used to sync USB timings by receiving a SOF event. Same goes for the SAI driver. 

Currently the PWM driver holds the complete timer instance, and a driver couldn't be written without kernel modification.

I was not planning to modify the PWM api, as the way Linux implements several features for the timer driver (capture, encoder, hall). But I was stuck using the current implementation.

Suggestions are welcome.

henrikbrixandersen commented:
> Suggestions are welcome.

@maxvankessel I am currently implementing quadrature decoder, frequency counter and pulse width measurement support atop of the NXP FlexTimer. What I have done so far is to use different device tree bindings per function, but this too feel wrong (the devicetree should describe the hardware, not the software configuration of the hardware). I too have a bit of duplicated code between the various drivers utilizing the FlexTimer, but it's not too much, I think.

A common approach to this problem (one device instance, different drivers on top) within Zephyr would be nice. IMO, this should be taken into consideration in the current devicetree rework effort. /cc @galak.

maxvankessel committed:
drivers: mfd: add timer instance 5

instance 5 Seemed to have slip past

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

maxvankessel committed:
dts: stm32 timer: comply to new dts tags

Deprecation of the parent-bus and child-bus

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

maxvankessel committed:
drivers: mfd: change clock control include

Corrected clock control include

Signed-off-by: Max van Kessel <max_van_kessel@msn.com>

gmarull commented on drivers/mfd/mfd_timer_stm32.c:
```diff
@@ -0,0 +1,277 @@
+		LL_TIM_SetTriggerOutput(tim, mode);
+
+		/* TODO: create binding for me ?*/
+		LL_TIM_DisableARRPreload(tim);
```
That should be configurable as `TODO` suggests

gmarull reviewed (commented):
@maxvankessel Overall looks like a nice addition, as it would allow to implement more specific drivers on top of STM32 timer generic functionality. I have a custom driver that would benefit from a these changes (I'm using a master & slave configuration). Let me know if I can help on testing or code changes.

gmarull commented on drivers/mfd/mfd_timer_stm32.c:
```diff
@@ -0,0 +1,277 @@
+			LL_TIM_SetTriggerInput(tim, mode);
+		}
+
+		mode = cfg->master_trig << TIM_CR2_MMS_Pos;
```
Example of the mentioned idea: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/uart_stm32.c#L139
